main/uart_nic.c

```diff
@@ -403,37 +514,115 @@ static void IRAM_ATTR check_online_status() {
     }
 }
 
+static void IRAM_ATTR clear_stored_ssids() {
+    memset(scan.stored_ssids, 0, scan.stored_ssids_count * sizeof(ScanResult));
+    scan.stored_ssids_count = 0;
+}
+
+static void IRAM_ATTR store_scanned_ssids(wifi_ap_record_t *aps, int ap_count) {
+    for (uint8_t i = 0; i < ap_count; ++i) {
+        bool found = false;
+        if (scan.stored_ssids_count == SCAN_MAX_STORED_SSIDS) {
+            break;
+        }
+        for (uint8_t j = 0; j < scan.stored_ssids_count; ++j) {
+            ESP_LOGI(TAG, "Comparing >%s< and %s", (char *)scan.stored_ssids[j].ssid, (char *)aps[i].ssid);
+            if (memcmp(scan.stored_ssids[j].ssid, aps[i].ssid, SSID_LEN) == 0) {
+                found = true;
+                break;
+            }
+        }
+        if (!found) {
+            memcpy(scan.stored_ssids[scan.stored_ssids_count].ssid, aps[i].ssid, SSID_LEN);
+            scan.stored_ssids[scan.stored_ssids_count].needs_password = aps[i].authmode != WIFI_AUTH_OPEN;
+            scan.stored_ssids_count ++;
+            ESP_LOGI(TAG, "Found SSID: %s", aps[i].ssid);
+        }
+    }
+
+    struct uart0_tx_queue_item queue_item = {0};
+    queue_item.header.type = MSG_SCAN_AP_CNT;
+    queue_item.header.ap_count = scan.stored_ssids_count;
+    if (xQueueSendToBack(uart0_tx_queue, &queue_item, 0) != pdTRUE) {
+        ESP_LOGE(TAG, "xQueueSendToBack failed (uart0tx)");
+    }
+    ESP_LOGI(TAG, "Scan done. Found: %d", scan.stored_ssids_count);
+}
+
+static void IRAM_ATTR send_scanned_ssid(uint8_t index) {
+    struct uart0_tx_queue_item queue_item = {0};
+    queue_item.header.type = MSG_SCAN_AP_GET;
+    queue_item.header.size = htons(sizeof(scan.stored_ssids[0]));
+    queue_item.data = (uint8_t *)(&scan.stored_ssids[index]);
+
+    if (xQueueSendToBack(uart0_tx_queue, &queue_item, 0) != pdTRUE) {
+        ESP_LOGE(TAG, "xQueueSendToBack failed (uart0tx)");
+    }
+}
+
+static void IRAM_ATTR handle_rx_msg_scan_start(uint8_t* data, struct header header) {
+    clear_stored_ssids();
+    ESP_LOGI(TAG, "Starting scan...");
+
+    start_wifi_scan(&store_scanned_ssids, SCAN_TYPE_LOOPING_SHORT);
+}
+
+static void IRAM_ATTR handle_rx_msg_scan_stop(uint8_t* data, struct header header) {
+    // Response is send automatically after scan is stopped
+    scan.in_progress = false;
+    scan.scan_type = SCAN_TYPE_UNKNOWN;
+    ESP_ERROR_CHECK(esp_wifi_scan_stop());
+}
+
+static void IRAM_ATTR handle_rx_msg_scan_get(uint8_t* data, struct header header) {
+    send_scanned_ssid(header.ap_index);
```

Comment:
Should there be some kind of „range check“ somewhere?

Reply:
I would prefer that it would happen on the printer side since the protocol doesn't support sending errors. I can modulo the index (but need to zero out the data that you object about to be able to return zeroth element if none are found) or I can "silently" fail (I can log an error, but nobody will know about it until they connect to the error pin with something that can read it).

Reply:
In these areas, I prefer to be somewhat defensive. I fully agree the primary check should be on the printer side and I'm fine classifying such event protocol violation / printer fault / bug.

I'm merely proposing to limit the „blast radius“ of such bug ‒ silently not answering, or providing some kind of null/empty answer instead of crashing the ESP firmware or sending some (potentially secret data) over the wire.

In a sense, it's a separate firmware running on a separate chip and it should do some (at least minimal) diligence against external communications ‒ that is, not expect whatever comes to be correct and non-malicious.

Anyway, any kind of minimal check and handling would be sufficient.